geode-core/src/test/java/org/apache/geode/internal/cache/InternalCacheBuilderTest.java

```diff
@@ -0,0 +1,615 @@
+
+import java.util.Properties;
+import java.util.function.Supplier;
+
+import org.assertj.core.api.ThrowableAssert.ThrowingCallable;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mock;
+
+import org.apache.geode.cache.CacheExistsException;
+import org.apache.geode.distributed.internal.DistributionConfig;
+import org.apache.geode.distributed.internal.InternalDistributedSystem;
+import org.apache.geode.internal.cache.InternalCacheBuilder.InternalCacheConstructor;
+import org.apache.geode.internal.cache.InternalCacheBuilder.InternalDistributedSystemConstructor;
+
+public class InternalCacheBuilderTest {
+
+  private static final int ANY_SYSTEM_ID = 12;
+  private static final String ANY_MEMBER_NAME = "a-member-name";
+
+  private static final Supplier<InternalDistributedSystem> THROWING_SYSTEM_SUPPLIER =
+      () -> {
+        throw new AssertionError("throwing system supplier");
+      };
+  private static final Supplier<InternalCache> THROWING_CACHE_SUPPLIER =
+      () -> {
+        throw new AssertionError("throwing cache supplier");
+      };
+
+  private static final InternalDistributedSystemConstructor THROWING_SYSTEM_CONSTRUCTOR =
+      (a, b) -> {
+        throw new AssertionError("throwing system constructor");
+      };
+  private static final InternalCacheConstructor THROWING_CACHE_CONSTRUCTOR =
+      (a, b, c, d, e, f) -> {
+        throw new AssertionError("throwing cache constructor");
+      };
+
+  @Mock
+  private Supplier<InternalDistributedSystem> nullSingletonSystemSupplier;
+
+  @Mock
+  private Supplier<InternalCache> nullSingletonCacheSupplier;
+
+  @Before
+  public void setUp() {
+    initMocks(this);
+
+    when(nullSingletonSystemSupplier.get()).thenReturn(null);
+    when(nullSingletonCacheSupplier.get()).thenReturn(null);
+
+    InternalDistributedSystem.ALLOW_MULTIPLE_SYSTEMS = false;
+  }
+
+  @Test // null Properties
+  public void create_throwsNullPointerException_ifConfigPropertiesIsNull() {
+    InternalCacheBuilder internalCacheBuilder = new InternalCacheBuilder(
+        null, new CacheConfig(),
+        nullSingletonSystemSupplier, constructorOf(constructedSystem()),
+        nullSingletonCacheSupplier, constructorOf(constructedCache()));
+
+    Throwable thrown = catchThrowable(() -> internalCacheBuilder
+        .create());
+
+    assertThat(thrown).isInstanceOf(NullPointerException.class);
+  }
+
+  @Test // null CacheConfig
+  public void create_throwsNullPointerException_andCacheConfigIsNull() {
+    InternalCacheBuilder internalCacheBuilder = new InternalCacheBuilder(
+        new Properties(), null,
+        nullSingletonSystemSupplier, constructorOf(constructedSystem()),
+        nullSingletonCacheSupplier, constructorOf(constructedCache()));
+
+    Throwable thrown = catchThrowable(() -> internalCacheBuilder
+        .create());
+
+    assertThat(thrown).isInstanceOf(NullPointerException.class);
+  }
+
+  @Test // scenario 1 - no system/no cache - constructs system
```

Comment:
I think most of the comments in this test file are covered by the test name itself.

Reply:
These scenario #s match up with #s in a spreadsheet that we created to make sure we weren't missing any tests. I'm planning to delete them before merging to develop.